sdk/src/azure/core/az_json_token.c

```diff
@@ -333,7 +333,8 @@ AZ_NODISCARD az_result az_json_string_unescape(
 {
   _az_PRECONDITION_VALID_SPAN(json_string, 1, false);
   _az_PRECONDITION_NOT_NULL(destination);
-  _az_PRECONDITION(destination_max_size > 0);
+  // The destination needs to be larger than the input, for null terminator.
+  _az_PRECONDITION(destination_max_size > az_span_size(json_string));
 
   int32_t position = 0;
   int32_t span_size = az_span_size(json_string);
```

Comment:
Since size is precondition, `if (position > destination_max_size)` can be dropped.
It checks on every loop iteration.

Either way, even if we keep it, it would be much cheaper to just test it before the loop, just as you do with precondition.

Then, it could be a single `if (current_char == '\\') { if (i < span_size - 1) { ... } else { ... } }`, instead of checking `if (current_char == '\\')` twice (even if not for optimization, but readability as well).

Reply:
You are right.

I think I am going to re-look at this whole implementation, including the optimizations you mentioned, as part of:
https://github.com/Azure/azure-sdk-for-c/issues/2302

I want to keep the code change in this PR minimal, for now, to avoid too much churn, since it will likely be throw away work.